Mortar mechanical contact constraints for dynamics

recuero opened the pull request:
Refs. #19671

recuero committed:
Add initial dynamic contact constraints and example.

 - Tested with normal contact and interpolate_normals = false
 - Something is off at the edge of the patch

recuero committed:
Latest changes (#19671)

Attempt at populating AutomaticMortarGeneration with data from the mortar contact constraint (weighted gap)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/921568/) on b2f7041 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c65ead690c99f7dd907692b246bf4af004149018`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/921602/) on 77f8dec wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c65ead690c99f7dd907692b246bf4af004149018`

recuero committed:
Add auxiliary kernel to compute kinetic energy (#19671)

lindsayad committed:
Create old weighted gap member and populate in timestepSetup

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/922029/) on aafe200 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1c4a0ba58e8a10e4900a0bc6dbea0ea973530207`

recuero committed:
WIP Archard's law auxiliary kernel/'dynamic' creep (#19671)

recuero committed:
Stabilization of mortar contact constraints for dynamics (#19671)

The nodal old weighted gap is used to trigger the enforcement of
the persistency condition (\dot(\tilde{gap}) * _dt) once nodes come
into contact. Examples show that contact forces are properly
stabilized. Small interpenetration may occur due to the time
discretization. The added parameter `capture_tolerance` can be
tuned to define the relative distance used to consider that a node
was "sufficiently" in contact at the previous step.

Also doing some clean up in this commit: Remove methods from
AutomaticMortarGeneration no longer needed and remove
and auxiliary kernel that had no purpose.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/922450/) on cf87f56 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 626704c07631478493459c280339c4f7045337e7`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/922487/) on 9d75ba4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19675/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/922679/) on 0d9ba0b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 626704c07631478493459c280339c4f7045337e7`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/922773/) on 14288d9 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>626704</th>
      <th colspan=4>#19675 14288d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.18%</td>
      <td>82.19%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72924</td>
      <td>72925</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15808</td>
      <td>15807</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19675/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19675/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>626704</th>
      <th colspan=4>#19675 14288d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.49%</td>
      <td>85.86%</td>
      <td>-2.63%</td>
      <td>58.70%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2521</td>
      <td>2683</td>
      <td>+162</td>
      <td>162</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>328</td>
      <td>442</td>
      <td>+114</td>
      <td>114</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19675/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19675/coverage/contact)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>626704</th>
      <th colspan=4>#19675 14288d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.21%</td>
      <td>89.22%</td>
      <td>+0.00%</td>
      <td>95.45%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>25203</td>
      <td>25224</td>
      <td>+21</td>
      <td>21</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3048</td>
      <td>3049</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19675/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19675/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19675/coverage/xfem)

</details>

## Warnings

- `contact` new line coverage rate 58.70% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/922773/) on 14288d9 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.86% is less than the required 87.0%

loganharbour commented:
@lindsayad @dschwen one of you assign yourselves please - don't know who is more appropriate.

recuero commented:
@recuero

recuero committed:
Fix bad merge

recuero committed:
Include Newmark-beta in the dynamic mortar contact constraints (#19671)

The stabilization of the normal contact constraint requires adding the
time discretization integration scheme in the constraint. This need is
obvious when running a challenging problem such as the ironing problem
when large deformation and nodal element passage come into play.

For now, we only 'allow' Newmark-beta integrator, which is fairly
standard for solid mechanics. The alpha parameter of HHT can be added
in the future.

recuero committed:
Newmark parameters required for mortar contact dynamics (#19671)

Also add a creep test and update existing ones.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/945552/) on 9a132a3 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.84% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/945552/) on 9a132a3 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.84% is less than the required 87.0%

recuero committed:
Add mortar auxiliary to compute weighted velocity and Archard's wear

Mortar auxiliary kernels can compute incremental quantities.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947124/) on 7353cb2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 752b49c85cdc75523552b245c55005fd88010d9a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/947376/) on 4b14b34 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 752b49c85cdc75523552b245c55005fd88010d9a`

recuero commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -84,7 +90,12 @@ MortarNodalAuxKernelTempl<ComputeValueType>::compute()
+  if (!_incremental)
+    _var.setNodalValue(value / total_volume);
+  else
+    _var.setNodalValue(value / total_volume + _u_old[0]);
```
Am I indexing `_u_old` with `[0]` correctly here (since it's always a nodal variable)?

lindsayad replied:
Yes. But you can add an assertion here that `_u_old.size() == 1` just to be sure

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/948249/) on 4795917 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 85.53% is less than the required 87.0%

lindsayad committed:
Compute each mortar nodal aux variable in its own loop

Then we can be sure that we won't erase the aux var we just computed.
This will allow us to also do dependency ordering for these aux kernels

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/949712/) on 4ea8659 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2be69d0465629dc8068934b17e5f079b4b3630f3`

recuero committed:
Cleanup and add 3d frictional test with dynamics (#19671).

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/950038/) on 7b18097 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 86.67% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/950161/) on 33121cb wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 86.67% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/950246/) on 3b4cdc2 wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 86.67% is less than the required 87.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/951044/) on 6cbe8be wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 86.73% is less than the required 87.0%

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/951573/) on 3734db6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19675/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19675/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format db314bead5d74ab94880069dfbcd0967183f2805`

dschwen commented on framework/include/auxkernels/AuxKernel.h:
```diff
@@ -94,6 +94,11 @@ class AuxKernelTempl : public MooseObject,
   bool isNodal() const { return _nodal; }
 
+  /**
+   * @return whether this is a mortar auxiliary kernel
```
what is a "mortar auxiliary kernel"?

dschwen replied:
It would be a bit clearer if this would use `dynamic_cast<MortarNodalAuxKernelTempl<is_ad>*>(this) != nullptr`

dschwen replied:
or if this method didn't exist and said dynamic cast was made inside the framework

lindsayad replied:
@recuero I am fine with this suggested change. Then we can make the method non-virtual

recuero replied:
This should be addressed now.

dschwen commented on modules/contact/src/auxkernels/MortarArchardsLawAux.C:
```diff
@@ -0,0 +1,128 @@
+  params.addCoupledVar("v",
+                       "Optional variable to take the value of. If omitted the value of the "
+                       "`variable` itself is returned.");
+  params.addRequiredCoupledVar("disp_x", "The x displacement variable");
+  params.addRequiredCoupledVar("disp_y", "The y displacement variable");
+  params.addCoupledVar("disp_z", "The z displacement variable");
```
I think we've been moving away from this pattern for a while, instead using `displacements` to couple multiple components (using `coupledValues` into a `std::vector<VariableValue *>`

recuero replied:
We need `MooseVariables` to have access to methods to obtain velocities and neighbor velocities, don't we?

dschwen commented on modules/contact/src/auxkernels/MortarArchardsLawAux.C:
```diff
@@ -0,0 +1,128 @@
+}
+
+void
+MortarArchardsLawAux::computeQpProperties()
```
weird to see `computeQpProperties` in an AuxKernel...

dschwen commented on modules/contact/src/auxkernels/WeightedGapVelAux.C:
```diff
@@ -0,0 +1,103 @@
+    _i(0),
+    _qp(0)
+{
+
```
```suggestion
```

recuero replied:
Should be addressed.

dschwen commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -0,0 +1,208 @@
+    _newmark_gamma(getParam<Real>("newmark_gamma"))
+{
+  mooseAssert(!_interpolate_normals,
+              "Dynamic mortar mechanical contact constraints require the surface geometry be "
```
```suggestion
              "Dynamic mortar mechanical contact constraints require the surface geometry to be "
```

recuero replied:
Should be addressed.

dschwen commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -0,0 +1,208 @@
+              "attached to nodes");
+
+  if (!useDual())
+    mooseError("Dynamic mortar contact constraints requires the use of Lagrange's multipliers dual "
```
```suggestion
    mooseError("Dynamic mortar contact constraints requires the use of Lagrange multiplier dual "
```

recuero replied:
Should be addressed.

dschwen reviewed (commented):
After reading the PR I think I know what a MortarNodalAuxKernel is, but it would be useful for other developers to have some doco on that class of AuxKernels.

recuero committed:
Address review comments

dschwen commented on modules/contact/src/auxkernels/MortarArchardsLawAux.C:
```diff
@@ -0,0 +1,128 @@
+    _friction_coefficient(getParam<Real>("friction_coefficient")),
+    _energy_wear_coefficient(getParam<Real>("energy_wear_coefficient")),
+    _has_disp_z(isCoupled("disp_z")),
+    _disp_x(*getVar("disp_x", 0)),
+    _disp_y(*getVar("disp_y", 0)),
+    _disp_z(getVar("disp_z", 0)),
```
```suggestion
    _displacements({getVar("displacements", 0)), getVar("displacements", 1), getVar("displacements", 2)}),
```

for example

lindsayad commented on modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C:
```diff
@@ -0,0 +1,371 @@
+  }
+  else
+  {
+    _assembly.cacheResidual(friction_dof_indices[0], dof_residual.value(), _vector_tags);
+    _assembly.cacheResidual(friction_dof_indices[1], dof_residual_dir.value(), _vector_tags);
```
I have a theory that this could be one of the reasons why automatic scaling convergence has always been so bad for you. `cacheResidual` does not apply the variable's scaling factor but `processDerivatives` does. There is a new API in `Assembly` called `processResidual` and it has the same argument ordering as `processDerivatives` (but takes `_vector_tags` instead of `_matrix_tags`). Let's switch to using `processResidual` here and then let's make sure that we do this everywhere in this PR ... and also in the non-dynamic classes. This is something that has only recently come to my attention and I haven't gotten around to fixing up contact ... because I forgot I use this design here

recuero replied:
Is `processResidual` in `FVInterfaceKernel`?

lindsayad replied:
There is a `FVInterfaceKernel::processResidual` but you want to call `_assembly.processResidual`

recuero replied:
Along the lines of what you explain above: Are the corresponding displacement residuals properly scaled? (I can't see where this is happening)

lindsayad replied:
A class like this is only computing the LM residual (well at least that's typically what the 'LM' in the class name has indicated). The displacement residuals, unless `computeResidual` and `computeJacobian` have been overridden, are taken care of in `ADMortarConstraint` and this happens behind the scenes after `accumulateTaggedLocalResidual`

recuero replied:
The displacement residuals are also taken care of now in `ADMortarLagrangeConstraint`. But this class follows the same design as `ADMortarConstraint`, including calls to `accumulateTaggedLocalResidual`.

Did some initial testing and didn't notice significant improvement with `automatic_scaling` using `processResidual`, but will keep looking.

lindsayad replied:
It's totally possible that unscaled will still fare better since the residuals will be scaled more "physically" and automatic scaling, by default, goes wholly by the Jacobian.

At least these changes should make results of `-snes_test_jacobian` better when automatic scaling is present

recuero committed:
Switch to processResidual to cache residuals in mortar contact constraints.

This is suggested by Alex to effectively account for scaling (i.e. discontinue the
 use of cacheResidual).

(#19671)

recuero commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -326,6 +326,11 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+  /**
+   *  Populate node to weighted gap map.
+   */
+  void setNodalWeightedGapMap(const Node * node, const Real weighted_gap);
```
It's not necessary. Removed it.

recuero committed:
Remove fixed scaling from mechanical LMs in thermomechanical tests.

These tests started taking more time once suboptimal scaling for contact
pressure LM is accounted for via the processResidual assembly method.

lindsayad commented:
So those tests started running worse after our "fix"? Can we check to see that the Jacobian improved?

recuero commented:
Yeah, they started running worse as the existing manual scaling factor started being taken into account.

For one of these models, the Jacobian difference ` ||J - Jfd||_F` improves as a more appropriate scaling factor is used on the LM variable manually. This improvement is clearly translated into better convergence behavior.

I converted these tests to use the weighted gap object and remember spending some time on the scaling and being a bit confused about the results. Now I think I can make sense out of the results for the fixed variable scaling.

lindsayad commented:
@recuero merge when ready or maybe you're waiting on another look from @dschwen ?
